### PR TITLE
INT-3685: Introduce STOMP Adapters

### PR DESCRIPTION
**Pull request by artembilan**
JIRA: https://jira.spring.io/browse/INT-3685
- Add `StompInboundChannelAdapter` to subscribe to STOMP destination and receive messages from them.
  Destinations can be added/removed (hence subscribed/unsubscribed) at runtime.
- Add `stompMessageHandler` to the send messages to STOMP destinations.
- `destination` can be extracted from `MessageHeaders`, the `destination` and `destinationExpression` are also supported.
- `RECEIPT` Frame is also supported and emitted as `StompReceiptEvent`
- `StompExceptionEvent` is emitted when we have a `failure` for `CONNECT` Frame or as a reaction to the `ERROR` Frame.
- Introduce `StompSessionManager` abstraction to manage the single `StompSession` and allow to share it between different adapters.
- Add `WebSocketStompSessionManager` implementation over `WebSocketStompClient`
- Add `StompHeaderMapper`

**TODO**
- Namespace support
- Docs (can be done in separate JIRA)

CC. @rstoyanchev

### CODE REVIEW

**Comment by artembilan** [comment]
Correction: no need to worry about Reactor 1.1 stuff. Spring Framework has been upgraded to Reactor 2 recently: https://github.com/spring-projects/spring-framework/commit/60b19c784d7c8431ac0545a18607fe18fff4bfda

**Comment by artembilan** [comment]
Squeshed after push for `Reactor2TcpStompClient` and tests against embedded STOMP server.

File: spring-integration-core/src/main/java/org/springframework/integration/endpoint/MessageProducerSupport.java [hunk 1]
```diff
@@ -66,6 +66,10 @@ public void setShouldTrack(boolean shouldTrack) {
 		this.shouldTrack = shouldTrack;
 	}
 
+	public MessagingTemplate getMessagingTemplate() {
+		return messagingTemplate;
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/endpoint/MessageProducerSupport.java]: What is the purpose of this? I see no references.

File: spring-integration-core/src/main/java/org/springframework/integration/endpoint/MessageProducerSupport.java [hunk 2]
```diff
@@ -66,6 +66,10 @@ public void setShouldTrack(boolean shouldTrack) {
 		this.shouldTrack = shouldTrack;
 	}
 
+	public MessagingTemplate getMessagingTemplate() {
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/endpoint/MessageProducerSupport.java]: protected?

File: spring-integration-stomp/src/main/java/org/springframework/integration/stomp/AbstractStompSessionManager.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+				@Override
+				public void onFailure(Throwable e) {
+					logger.error("STOMP connect error.", e);
+					if (applicationEventPublisher != null) {
+						applicationEventPublisher.publishEvent(
+								new StompExceptionEvent(AbstractStompSessionManager.this, e));
+					}
+					AbstractStompSessionManager.this.stompSession = null;
+					AbstractStompSessionManager.this.stompSessionListenableFuture = null;
```
Comment (garyrussell) [inline spring-integration-stomp/src/main/java/org/springframework/integration/stomp/AbstractStompSessionManager.java]: This looks a bit smelly - won't deal with concurrent `connect` calls and the `start()` methods have no protection against that.

File: spring-integration-stomp/src/main/java/org/springframework/integration/stomp/inbound/StompInboundChannelAdapter.java [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+			for (String d : destination) {
+				if (this.destinations.remove(d)) {
+					if (this.logger.isDebugEnabled()) {
+						logger.debug("Removed '" + d + "' from subscriptions.");
+					}
+					StompSession.Subscription subscription = this.subscriptions.get(d);
+					if (subscription != null) {
+						subscription.unsubscribe();
+					}
+				}
```
Comment (garyrussell) [inline spring-integration-stomp/src/main/java/org/springframework/integration/stomp/inbound/StompInboundChannelAdapter.java]: debug log? (tried to remove non-existent subs)

File: spring-integration-stomp/src/main/java/org/springframework/integration/stomp/support/StompHeaderMapper.java [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+	@Override
+	public Map<String, Object> toHeaders(StompHeaders source) {
+		Map<String, Object> target = new HashMap<String, Object>();
+		for (String name : source.keySet()) {
+			if (shouldMapHeader(name, this.inboundHeaderNames)) {
+				if (StompHeaders.CONTENT_TYPE.equals(name)) {
+					target.put(MessageHeaders.CONTENT_TYPE, source.getContentType());
+				}
+				else {
+					target.put(name, source.getFirst(name));
```
Comment (garyrussell) [inline spring-integration-stomp/src/main/java/org/springframework/integration/stomp/support/StompHeaderMapper.java]: Why only the first value?

Also, shouldn't we be prefixing the headers `stomp_` ??

**Comment by garyrussell** [comment]
That's my quick first pass; I really need javadocs/docs for a full review.

File: spring-integration-stomp/src/main/java/org/springframework/integration/stomp/support/StompHeaderMapper.java (hunk 5 above)
Comment (artembilan) [inline spring-integration-stomp/src/main/java/org/springframework/integration/stomp/support/StompHeaderMapper.java]: https://stomp.github.io/stomp-specification-1.2.html#Repeated_Header_Entries:
> If a client or a server receives repeated frame header entries, only the first header entry SHOULD be used as the value of header entry.

And see the impl of the previous `source.getContentType()`.

Agreed regarding `stomp_`, although `StompDecoder` just build a `Message<byte[]>` with `StompHeaders` as is.
However further it looks like:
```
private void invokeHandler(StompFrameHandler handler, Message<byte[]> message, StompHeaders stompHeaders) {
	if (message.getPayload().length == 0) {
		handler.handleFrame(stompHeaders, null);
		return;
	}
	Type type = handler.getPayloadType(stompHeaders);
	Class<?> payloadType = ResolvableType.forType(type).getRawClass();
	Object object = getMessageConverter().fromMessage(message, payloadType);
	if (object == null) {
		throw new MessageConversionException("No suitable converter, payloadType=" + payloadType);
	}
	handler.handleFrame(stompHeaders, object);
}
```

So, I think some `PassThruMessageConverter` impl as injection to the `StompClientSupport` will help us to have a better throughput without that extra conversion `message -> payload + headers -> headerMapper + message`.
And with that we leave the target payload conversion to the end-user in the downstream flow.

Will address that in the next commit, if you don't have objections.
And, yes, with `stomp_`. 
How about some new `IntegrationStompHeaders` do not confuse end-users with `StompHeaders` from SF?
And I guess `StompHeaderMapper#fromHeaders`  should be tolerant and maps headers from both `StompHeaders` and `IntegrationStompHeaders`.

File: spring-integration-stomp/src/main/java/org/springframework/integration/stomp/inbound/StompInboundChannelAdapter.java (hunk 4 above)
Comment (artembilan) [inline spring-integration-stomp/src/main/java/org/springframework/integration/stomp/inbound/StompInboundChannelAdapter.java]: Not sure if it makes sense.
Just copied/pasted from the `MqttPahoMessageDrivenChannelAdapter`.
Or I just don't understand what you mean here...

**Comment by artembilan** [comment]
Pushed

**Comment by artembilan** [comment]
If you find it as fine, I'd prefer to address Namespace and Docs with separate JIRAs.
Although I can follow with your request anyway.

File: spring-integration-stomp/src/main/java/org/springframework/integration/stomp/inbound/StompInboundChannelAdapter.java (hunk 4 above)
Comment (garyrussell) [inline spring-integration-stomp/src/main/java/org/springframework/integration/stomp/inbound/StompInboundChannelAdapter.java]: I am just suggesting that if someone unsubscribes from a non-existent subscription we could log a debug message (and add to mqtt if it's missing there too).

Comment (artembilan) [inline spring-integration-stomp/src/main/java/org/springframework/integration/stomp/inbound/StompInboundChannelAdapter.java]: Pushed

**Comment by garyrussell** [comment]
Hmmm... seeing this in the gradle build...

    org.springframework.integration.stomp.outbound.StompMessageHandlerWebSocketIntegrationTests > testStompMessageHandler FAILED
    org.springframework.messaging.MessageDeliveryException at StompMessageHandlerWebSocketIntegrationTests.java:122
        Caused by: java.lang.ClassCastException at StompMessageHandlerWebSocketIntegrationTests.java:122

...but I don't see it in the IDE.

Investigating.

**Comment by rstoyanchev** [comment]
It's probably related to the change from StringMessageConverter to SimpleMessageConverter as the default converter used. https://github.com/spring-projects/spring-framework/commit/c48e8708a7cc3117603f843166a3389c7ad16a07

**Comment by garyrussell** [comment]
Cool; thanks @rstoyanchev - refreshing the eclipse deps give me the error in STS now.

**Comment by garyrussell** [comment]
Merged as 9e4c7e00db6f3f000d6709ba6385fcd2c669cb40

Polish: garyrussell/spring-integration@e466cca

File: spring-integration-stomp/src/main/java/org/springframework/integration/stomp/outbound/StompMessageHandler.java [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+		StompHeaders stompHeaders = new StompHeaders();
+		this.headerMapper.fromHeaders(message.getHeaders(), stompHeaders);
+		if (stompHeaders.getDestination() == null) {
+			Assert.state(this.destinationExpression != null, "One of 'destination' or 'destinationExpression' must be" +
+					" provided, if message header doesn't supply 'destination' STOMP header.");
+			String destination = this.destinationExpression.getValue(this.evaluationContext, message, String.class);
+			stompHeaders.setDestination(destination);
+		}
+
+		final StompSession.Receiptable receiptable = this.stompSession.send(stompHeaders, message.getPayload());
```
Comment (garyrussell) [inline spring-integration-stomp/src/main/java/org/springframework/integration/stomp/outbound/StompMessageHandler.java]: Need to wait for the `stompSession` to be established by the callback; maybe a latch set up by `start()` ?

Comment (garyrussell) [inline spring-integration-stomp/src/main/java/org/springframework/integration/stomp/outbound/StompMessageHandler.java]: Forget that; I see you set running in the callback.

File: spring-integration-stomp/src/main/java/org/springframework/integration/stomp/AbstractStompSessionManager.java [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+		this.stompSessionListenableFuture = doConnect(this.compositeStompSessionHandler);
+		this.stompSessionListenableFuture.addCallback(new ListenableFutureCallback<StompSession>() {
+
+			@Override
+			public void onFailure(Throwable e) {
+				logger.error("STOMP connect error.", e);
+				if (applicationEventPublisher != null) {
+					applicationEventPublisher.publishEvent(
+							new StompExceptionEvent(AbstractStompSessionManager.this, e));
+				}
```
Comment (garyrussell) [inline spring-integration-stomp/src/main/java/org/springframework/integration/stomp/AbstractStompSessionManager.java]: Shouldn't we schedule a new connect after a connect failure?

File: spring-integration-stomp/src/main/java/org/springframework/integration/stomp/outbound/StompMessageHandler.java [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+
+	@Override
+	public void start() {
+		this.stompSessionManager.connect(this.sessionHandler);
+	}
+
+	@Override
+	public void stop() {
+		this.stompSessionManager.disconnect(this.sessionHandler);
+		this.running = false;
```
Comment (garyrussell) [inline spring-integration-stomp/src/main/java/org/springframework/integration/stomp/outbound/StompMessageHandler.java]: Need to reset the boolean first.

File: spring-integration-stomp/src/main/java/org/springframework/integration/stomp/outbound/StompMessageHandler.java [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+					}
+				}
+
+			});
+		}
+	}
+
+	@Override
+	public void start() {
+		this.stompSessionManager.connect(this.sessionHandler);
```
Comment (garyrussell) [inline spring-integration-stomp/src/main/java/org/springframework/integration/stomp/outbound/StompMessageHandler.java]: Maybe we should make the start lazy? i.e. in `handleMessageInternal` we should try to connect if we're not connected?

Comment (artembilan) [inline spring-integration-stomp/src/main/java/org/springframework/integration/stomp/outbound/StompMessageHandler.java]: If you take a look to the `stompSessionManager.connect()` implementation, you'll see that it is just an *artificial* connect and we do only this:
```
public void connect(StompSessionHandler handler) {
	this.compositeStompSessionHandler.addHandler(handler);
}
```
So, any `connectIfNeeded()` doesn't make sense here. We establish `stompSession` here only within:
```
public void afterConnected(StompSession session, StompHeaders connectedHeaders) {
	StompMessageHandler.this.stompSession = session;
	StompMessageHandler.this.running = true;
}
```
Which would be possible only if the target `stompSessionManager` has been obtained the *real* connection.

File: spring-integration-stomp/src/main/java/org/springframework/integration/stomp/AbstractStompSessionManager.java (hunk 7 above)
Comment (artembilan) [inline spring-integration-stomp/src/main/java/org/springframework/integration/stomp/AbstractStompSessionManager.java]: I think I agree here and this `StompSessionManager` really should take care about reconnect to the lost connection for the session. Like you do in the `MqttPahoMessageDrivenChannelAdapter`

File: spring-integration-stomp/src/main/java/org/springframework/integration/stomp/outbound/StompMessageHandler.java (hunk 8 above)
Comment (artembilan) [inline spring-integration-stomp/src/main/java/org/springframework/integration/stomp/outbound/StompMessageHandler.java]: Accepted

**Pull request closed without merging**